### ISSUE TITLE
Add analyzer: Interlocked.CompareExchange when "value" argument is read from "location" argument

### ISSUE THREAD

**Issue by drewnoakes** [dotnet/runtime#45828]
Today I came across a property getter which was implemented as:

```c#
get
{
    return Interlocked.CompareExchange(ref this.someField, this.someField, null);
}
```

I can't think of a valid scenario where the `location` argument (first) should also be read for the `value` field.

This seems to be asking for trouble, as the read of the field for the second argument races with the cmpxchg.

My guess is that the author wanted to achieve what `Volatile.Read` offers.

**Comment by xtqqczze** [comment]
Perhaps the author wanted to be consistent by using the same synchronization primitive as elsewhere, and needed the latest value of `someField`;

Nevertheless, the code should probably have been written like this:

```csharp
get
{
    return Interlocked.CompareExchange(ref this.someField, null, null);
}
```

**Comment by drewnoakes** [comment]
I believe the intention was to return the value of the field, rather than swap it out for null. Perhaps a codefix isn't possible here, but I do think the pattern looks suspicious enough to warrant a diagnostic.

**Comment by xtqqczze** [comment]
>> ```csharp
>> get
>> {
>>     return Interlocked.CompareExchange(ref this.someField, null, null);
>> }
>> ```


> I believe the intention was to return the value of the field, rather than swap it out for null. Perhaps a codefix isn't possible here, but I do think the pattern looks suspicious enough to warrant a diagnostic.

@drewnoakes In the case that `someField` is `null`, `CompareExchange` will compare `someField` to `null`, succeed, and replace it with `null`, and return `null`. Therefore the value of `someField` is not modified.

On 32 bit platforms `Volatile.Read(Int64)` is implemented using this pattern:

https://github.com/dotnet/runtime/blob/574e6259ab0d26cf97c0eca07c8b08948018b07b/src/libraries/System.Private.CoreLib/src/System/Threading/Volatile.cs#L91-L97

**Comment by drewnoakes** [comment]
> In the case that someField is null, CompareExchange will compare someField to null, succeed, and replace it with null, and return null. Therefore the value of someField is not modified.

Yes, and perhaps this could be a second diagnostic. There are two cases which don't make any sense to me:

1. `Interlocked.CompareExchange(ref F, F, A);` &mdash; the original scenario
1. `Interlocked.CompareExchange(ref F, A, A);` &mdash; the scenario you have highlighted

> On 32 bit platforms Volatile.Read(Int64) is implemented using this pattern:

That's only true for `long` which is a 64 bit value, and would be subject to tearing.
